source/common/router/req_header_formatter.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Interface for all types of header formatters used for custom request headers.
+ */
+class HeaderFormatter {
+public:
+  virtual ~HeaderFormatter() {}
+
+  virtual const std::string
+  format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterPtr;
```

Comment:
@RomanDzhabarov That did not work either. :(. I got the same error.

RequestHeaderParser RequestHeaderParser::parse(const Json::Object& config) {
 36   RequestHeaderParser request_header_parser;
 37   std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
 38       header_formatter_map;
 39   if (config.hasObject("request_headers_to_add")) {
 40     std::vector<Json::ObjectSharedPtr> request_headers =
 41         config.getObjectArray("request_headers_to_add");
 42     for (const Json::ObjectSharedPtr& header : request_headers) {
 43       if (!header->getString("key").empty() && !header->getString("value").empty()) {
 44         ENVOY_LOG(debug, "adding key {} to header formatter map", header->getString("key"));
 45         HeaderFormatterPtr header_formatter =
 46             RequestHeaderParser::parseInternal(header->getString("value"));
 47         request_header_parser.header_formatter_map_.insert(std::make_pair(Http::LowerCaseString(header->getString("key")),std::move(header_formatter)));
 48       }
 49     }
 50   }
 51
 52   return request_header_parser;


error:

/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/ext/new_allocator.h:120:23: error: call to implicitly-deleted copy constructor of 'std::pair<const Envoy::Http::LowerCaseString, std::unique_ptr<Envoy::Router::HeaderFormatter, std::default_delete<Envoy::Router::HeaderFormatter> > >'
        { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }